src/esp/gfx/Shader.cpp

```diff
@@ -20,16 +20,18 @@ namespace gfx {
 
 namespace {
 // TODO: don't hardcode this
-std::unique_ptr<Magnum::GL::AbstractShaderProgram> shaderProgramFactory(
+std::shared_ptr<Magnum::GL::AbstractShaderProgram> shaderProgramFactory(
     ShaderType type,
     const ShaderConfiguration& cfg) {
   switch (type) {
     case ShaderType::INSTANCE_MESH_SHADER: {
-      return std::make_unique<gfx::PrimitiveIDShader>();
+      static auto primitiveShader = std::make_shared<gfx::PrimitiveIDShader>();
```

Comment:
Not sure about the `static` here -- this will mean the instance survives even after the GL context is destroyed, and then when creating a new GL context, this will still return the stale, no longer valid instance.

Same with multithreading -- in case someone would want to run multiple separate simulator instances on separate threads, this makes that impossible.

Reply:
Great catch, instead we will have to instantiate the shader factory as a member of the simulator, and inject it into this shader class, to ensure instances are torn down correctly, and are always unique across simulator instances.

Reply:
This can't stay unfortunately.  It will break things immediately.  Tearing down and re-spinning up the simulator is a very common occurrence.

Reply:
removed, I'll implement shader reuse later